harness/determined/layers/_workload_sequencer.py

```diff
@@ -200,12 +201,20 @@ def train(self, num_batches: int, op: core.SearcherOperation) -> WorkloadGenerat
         metrics = response.get("metrics", {}).get("avg_metrics", {})
         batch_metrics = response.get("metrics", {}).get("batch_metrics", [])
 
+        if self.records_per_epoch > 0:
+            epoch = math.ceil(
+                self.state.steps_completed * self.global_batch_size / self.records_per_epoch
+            )
+        else:
+            epoch = 0
+
```

Comment:
Calculating epoch from `records_per_epoch` is a bad idea, and it will frequently be wrong.  For instance, real-life datasets aren't always the same length every epoch, and every batch does not contain `global_batch_size` records.

We added `records_per_epoch` once upon a time when the workload sequencer existed only in the master, and there was no way to access the real dataset to figure out what epoch we were on.  It was messy but it was the best we could do at the time.

Now that we have moved the wlsq into the python code, new epoch-related features should be developed using actual epoch counts, rather than depending on calculated epoch derived from `records_per_epoch`.

Reply:
(so you would have to report epoch as a metric from each of EstimatorTrial, TFKerasTrial, and PyTorchTrial)